### DISCUSSION TITLE
Frontmatter manipulation like toggle

### DISCUSSION THREAD

**Discussion by ZeFish** [chrisgurney/obsidian-note-toolbar#30] [Scripting]
This is an amazing plugin with unlimited potential! I love it! Do you think it would be possible to use it as a way to manipulate frontmatter data? I guess I could create a template to add properties, but what about removing/editing or toggling properties?

i could see a lot of use case for that like the draft status, publish status and so on

**Comment by chrisgurney** [comment]
@ZeFish Hello! Thanks for the suggestion.

Have you looked at other plugins that might expose custom commands to do this? I personally haven't found anything yet -- as I look at plugins like [MetaEdit](https://github.com/chhoumann/MetaEdit) or [Meta Bind](https://github.com/mProjectsCode/obsidian-meta-bind-plugin), neither provide commands for single actions.

Meta Bind is close to what you want, creating buttons that can adjust properties, but buttons must be within notes I believe.

I might consider adding support for something like this, but I think this might be a bit of a rabbit hole of requests for rules and whatnot. 😅 I'll definitely think about it, and post updates here.

What kind of operations on properties would you like to see? What are some examples you might use this for?

**Comment by ZeFish** [comment]
I'll check, I also haven't found anything that can work with the frontmatter and I'm having hard time figuring out why.
Thanks for the suggestion of the two plugin but from what I see
* MetaEdit -> Seems even more work to update a frontmatter. For now it's seems simpler (for me) to
      1 - Get the property sidebar always shown
       2 - Switch to source mode and edit
* Meta Bind -> Frontmatter option (to me) is used as metadata (like publishing status) so I'm not looking for any more ui element in my notes. What I like about your plugin is the idea of the minimalistic ui

But here's a scenario, inspired by [Highlight Public Notes](https://github.com/dennisseidel/highlightpublicnotes-obsidian-plugin), it would be amazing to have a toggle button that could change it's border color based on the property.
I can Imagine a button wich is a dropdown to add/remove a list of tags. Could be helpfull for tag based organisation of notes.


Altho really a complexe plugin [Linter](https://github.com/platers/obsidian-linter) seems to have what it needs to modify the frontmatter data. 
[Update time on edit](https://github.com/beaussan/update-time-on-edit-obsidian)
[Yaml Manager](https://github.com/lijyze/obsidian-state-switcher)

Another use case is that I create a couple of theme that are displayed based on some cssclasses in my note. That way I can have multiple theme and use them based on the note.

I'll keep an eye out in the code to see if I can spot something interesting.

**Reply by chrisgurney** [reply]
Thanks for expanding on your use cases!

Interesting that you mentioned drop-down lists: sub-menu support is something I have in the back of my mind as well.

**Comment by FelipeRearden** [comment]
> 

@ZeFish  Sharing one idea…

I can manipulate frontmatter using a Templater script. 

This site has a lot of complex templater scripts 
https://zachyoung.dev/posts/templater-snippets

Maybe this one is what you are looking for…
![IMG_1163](https://github.com/chrisgurney/obsidian-note-toolbar/assets/85180696/72436fa4-8ca7-4c98-9f8a-c62d11886684)

If search on Discord you will see more code snippets that might help you 🙏

**Comment by FeralFlora** [comment]
See also #79

**Comment by chrisgurney** [comment]
@ZeFish I haven't forgotten about this. As @FelipeRearden hinted at, I'm thinking that having support for Templater scripts (in addition to Dataview scripts) might be a more general, flexible solution for your request.

I'm not 100% how this would work quite yet, and I'd love to see examples of other plugins support Templater/Dataview, so I could get some ideas.

If a toolbar button could execute a Templater script, do you think this would solve your particular use case?

**Reply by FelipeRearden** [reply]
> If a toolbar button could execute a Templater script

@chrisgurney just to share this information with you. 

Templater is able to transform it templater scripts into obsidian commands (which will appear into the Command Palette). This is a templater feature and this is the way that I use templater scripts with QuickAdd - as obsidian commands. 

My case, I run templater scripts from Note Toolbar using the approach described above without any issue. 

I am not a Data View user **yet**, this way I cant tell you about it.

**Reply by chrisgurney** [reply]
@FelipeRearden Interesting! I'll have to look into how that works. I think it would be useful to add a page to the Note Toolbar wiki/docs about how to get the command for a script!

**Comment by chrisgurney** [comment]
@ZeFish @FeralFlora Scripting support (Templater, Dataview, JavaScript via JS Engine) is now available in [the beta](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/171), which may help you with these use cases.

**Comment by chrisgurney** [comment]
(Just to close the loop on this discussion:)

**Create a JavaScript item**, select **Evaluate JavaScript**, and use the [Note Toolbar API](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Note-Toolbar-API) to get properties...

```ts
const createdDate = ntb.getProperty('created');
```

...and to set properties:

```ts
await ntb.setProperty('Created', moment().format('YYYY-MM-DD'));
await ntb.setProperty('cssclasses', 'myclass');
await ntb.setProperty('A Link', '[[Some Note]]');
await ntb.setProperty('A Number', 1234);
await ntb.setProperty('A List', ['asdf', 'asdf2']);
await ntb.setProperty('Delete Me', null); 
```
